### PR TITLE
Bind the SSL certificate, when presented with one.

### PR DESCRIPTION
**Pull request by bartonjs**

### CODE REVIEW

**Comment by bartonjs** [comment]
cc: @stephentoub, @vijaykota, @rajansingh10, @shrutigarg, @pgavlin , @CIPop , @davidsh, @SidharthNabar

**Comment by stephentoub** [comment]
LGTM

**Comment by bartonjs** [comment]
Ugh, the P2P build error that I saw previously is back; and doesn't manifest locally (or, bizarrely, when building the OpenSSL library by itself).

Anyone have any good ideas?

**Comment by bartonjs** [comment]
@weshaggard or @mmitche Any idea why S.S.C.OpenSsl builds, S.S.C.X509(S.S.C.OpenSsl) builds, but System.Net.Security(S.S.C.OpenSsl) doesn't build?

File: src/Common/src/Interop/Unix/libssl/Interop.libssl.cs [hunk 1]
```diff
@@ -121,10 +123,10 @@ internal static partial class libssl
         internal static extern unsafe IntPtr d2i_X509(IntPtr certPtr, ref byte* bufPtr, int len);
 
         [DllImport(Interop.Libraries.LibSsl, EntryPoint = "SSL_CTX_use_certificate")]
-        internal static extern int SSL_CTX_use_certificate(IntPtr ssl, IntPtr certPtr);
+        internal static extern int SSL_CTX_use_certificate(IntPtr ssl, SafeX509Handle certPtr);
 
         [DllImport(Interop.Libraries.LibSsl, EntryPoint = "SSL_CTX_use_PrivateKey")]
-        internal static extern int SSL_CTX_use_PrivateKey(IntPtr ssl, IntPtr certPtr);
+        internal static extern int SSL_CTX_use_PrivateKey(IntPtr ssl, SafeEvpPKeyHandle certPtr);
```
Comment (vijaykota) [inline src/Common/src/Interop/Unix/libssl/Interop.libssl.cs]: Nit: Existing issue: Rename parameter to keyPtr

File: src/System.Net.Security/src/System/Net/Unix/SecuritySafeHandles.cs [hunk 2]
```diff
@@ -67,10 +74,31 @@ internal SslProtocols Protocols
         public SafeFreeCredentials(X509Certificate certificate, SslProtocols protocols, EncryptionPolicy policy)
             : base(IntPtr.Zero, true)
         {
-            if (null != certificate)
+            Debug.Assert(
+                certificate == null || certificate is X509Certificate2,
+                "Only X509Certificate2 certificates are supported at this time");
+
+            X509Certificate2 cert = (X509Certificate2)certificate;
```
Comment (vijaykota) [inline src/System.Net.Security/src/System/Net/Unix/SecuritySafeHandles.cs]: Should we assert for HasPrivateKey as well?

**Comment by vijaykota** [comment]
LGTM

File: src/Common/src/Interop/Unix/libssl/Interop.libssl.cs (hunk 1 above)
Comment (bartonjs) [inline src/Common/src/Interop/Unix/libssl/Interop.libssl.cs]: Fixed (while rebasing for merge conflict).

File: src/System.Net.Security/src/System/Net/Unix/SecuritySafeHandles.cs (hunk 2 above)
Comment (bartonjs) [inline src/System.Net.Security/src/System/Net/Unix/SecuritySafeHandles.cs]: Done (while rebasing for merge conflict).

**Comment by bartonjs** [comment]
The previous failed build (the mysterious error of System.Security.Cryptography.OpenSsl.dll fails to build when referenced by System.Net.Security) was at http://dotnet-ci.cloudapp.net/job/dotnet_corefx_prtest/4336/; in case anyone can help try to figure out what's going on.

**Comment by mmitche** [comment]
@bartonjs No idea

**Comment by stephentoub** [comment]
@dotnet-bot test this please (after merging some changes from dev branch)

**Pull request merged**